spec/requests/hashtags_controller_spec.rb

```diff
@@ -14,134 +16,409 @@
+              },
+              "tags" => {
+                tag.name => tag.full_url,
+              },
+            )
+          end
+
+          it "does not return restricted categories or hidden tags" do
+            get "/hashtags.json", params: { slugs: [private_category.slug, hidden_tag.name] }
+
+            expect(response.status).to eq(200)
+            expect(response.parsed_body).to eq("categories" => {}, "tags" => {})
+          end
         end
 
-        it "does not return restricted categories or hidden tags" do
-          get "/hashtags.json", params: { slugs: [private_category.slug, hidden_tag.name] }
+        context "as admin" do
+          fab!(:admin) { Fabricate(:admin) }
 
-          expect(response.status).to eq(200)
-          expect(response.parsed_body).to eq("categories" => {}, "tags" => {})
+          before { sign_in(admin) }
+
+          it "returns restricted categories and hidden tags" do
+            group.add(admin)
+
+            get "/hashtags.json", params: { slugs: [private_category.slug, hidden_tag.name] }
+
+            expect(response.status).to eq(200)
+            expect(response.parsed_body).to eq(
+              "categories" => {
+                private_category.slug => private_category.url,
+              },
+              "tags" => {
+                hidden_tag.name => hidden_tag.full_url,
+              },
+            )
+          end
         end
-      end
 
-      context "as admin" do
-        fab!(:admin) { Fabricate(:admin) }
+        context "with sub-sub-categories" do
+          before do
+            SiteSetting.max_category_nesting = 3
+            sign_in(Fabricate(:user))
+          end
 
-        before { sign_in(admin) }
+          it "works" do
+            foo = Fabricate(:category_with_definition, slug: "foo")
+            foobar = Fabricate(:category_with_definition, slug: "bar", parent_category_id: foo.id)
+            foobarbaz =
+              Fabricate(:category_with_definition, slug: "baz", parent_category_id: foobar.id)
 
-        it "returns restricted categories and hidden tags" do
-          group.add(admin)
+            qux = Fabricate(:category_with_definition, slug: "qux")
+            quxbar = Fabricate(:category_with_definition, slug: "bar", parent_category_id: qux.id)
+            quxbarbaz =
+              Fabricate(:category_with_definition, slug: "baz", parent_category_id: quxbar.id)
 
-          get "/hashtags.json", params: { slugs: [private_category.slug, hidden_tag.name] }
+            get "/hashtags.json",
+                params: {
+                  slugs: [
+                    ":", # should not work
+                    "foo",
+                    "bar", # should not work
+                    "baz", # should not work
+                    "foo:bar",
+                    "bar:baz",
+                    "foo:bar:baz", # should not work
+                    "qux",
+                    "qux:bar",
+                    "qux:bar:baz", # should not work
```

Comment:
Wonder if these could be extracted and assigned with a name that explains why it "should not work", e.g. `invalid_slugs`, `existing_slugs`, and `nonexistent_slugs`, and then merged here.

Reply:
Yep good idea, will do